src/coprocessor_v2/plugin_registry.rs

```
@@ -2,28 +2,128 @@
 }
 
 #[allow(dead_code)]
 impl PluginRegistry {
     /// Creates a new `PluginRegistry`.
     pub fn new() -> Self {
-        PluginRegistry::default()
+        let registry = Arc::new(RwLock::new(PluginRegistryInner::default()));
+
+        PluginRegistry {
+            inner: registry,
+            fs_watcher: None,
+        }
+    }
+
+    /// Hot-reloads plugins from a given directory.
+    ///
+    /// All plugins that are already present in the directory will be loaded.
+    /// A background thread is spawned to watch file system events. If the library file of a loaded
+    /// plugin is deleted, the corresponding plugin is automatically unloaded; if a new library file
+    /// is placed into the directory, it will be automatically loaded into TiKV's coprocessor plugin
+    /// system.
+    ///
+    /// A file will only be loaded if it has the proper file ending of dynamic link libraries for
+    /// the current platform (`.so` for Linux, `.dylib` for MacOS, `.dll` for Windows).
+    pub fn start_hot_reloading(
+        &mut self,
+        plugin_directory: impl Into<PathBuf>,
+    ) -> notify::Result<()> {
+        let plugin_directory = plugin_directory.into();
+
+        // If this is the first call to `start_hot_reloading()`, create a new file system watcher
+        // and background thread for loading plugins. For later invocations, the same watcher and
+        // thread will be used.
+        if self.fs_watcher.is_none() {
+            let (tx, rx) = mpsc::channel();
+            let fs_watcher = notify::watcher(tx, Duration::from_secs(3)).unwrap();
+
+            let hot_reload_registry = self.inner.clone();
+            thread::spawn(move || {
+                // Simple helper functions for loading/unloading plugins.
+                let maybe_load = |file: &PathBuf| {
+                    if is_library_file(&file) {
+                        // Discard error.
+                        let _ = hot_reload_registry.write().unwrap().load_plugin(file);
```

Maybe better log it.

---

It's logged with `warn!`, but inside `load_plugin()`. Wasn't sure about where to place the logging. Is this okay?

---

alright, got it. If so, the convention to discard the error is to call `.ok()` on the result.